File: 1.1/py/1.1.py
import sys

sys.path.append("/home/adam/github_repos/CTCI/helpers")
from CTCI_helpers import run_tests
# Q: determine if a string contains all unique character
# Clarification:
#   -What character set are we using? If we have a fixed character set we can use 
#    an array instead of a dict/hashmap to store our already seen characters
#   -Should uppercase and lowercase be considered different characters?
#   -What about non-alphanumeric characters?

# Test Cases:
tests = {
    True : 
    [
        "abcdefgh",
        "oen luf"
    ],
    
    False : 
    [
        "repeat",
        "random letters"
    ]
}

# A1: Brute force solution
# Description: for each char in string, iterate through the rest of the string checking if they match 
# Time:  O(n^2)
# Space: O(1)

def isUnique(s):
    for i in range(len(s)):
        for j in range(i+1  , len(s)):
            if s[i] == s[j]:
                return False
    return True

run_tests(tests, isUnique, "brute force")
# A2: Fastest solution
# Description: Use a dict to store seen characters as we iterate through the array
# Time: O(n)
# Space: O(n)

def isUnique(s):
    seenChars = {}
    for char in s:
        if char in seenChars:
            return False
        seenChars[char] = True
    return True
run_tests(tests, isUnique, "fastest solution")

# A3: fastest solution using no additional space
# Description: If we sort the string first, then we only need to iterate through the string once afterward
# Time: O(nlog(n))
# Space: O(1)

def isUnique(s):
    s = sorted(s)
    i, j = 0, 1
    while j < len(s):
        if s[j] == s[i]:
            return False
        i += 1
        j += 1
    return True

run_tests(tests, isUnique, "O(1) space (sorted)")

File: 1.1/cpp/1.1.cpp
#include <iostream>
#include <functional>
#include <vector>
#include <map>
#include <typeinfo>
#include <algorithm>


// template <typename T, typename ...args> T call_func(std::function<T (args...)> func, args...){
//     return func(args...);
// }

// bool concat_params(int i, char c, std::string s){
//     std::cout << i << c << s << '\n';
//     return true;
// }

// brute force solution
bool isUnique1(std::string& s){
    for(int i = 0; i < s.length(); i++){
        for(int j = i + 1; j < s.length(); j++)
            if(s[i] == s[j]){
                return 0;
            }
    }   
    return 1; 
}

// using an array to store whether we have seen characters or not
bool isUnique2(std::string& s){
    std::array<bool, 256> charset {};

    for(int i = 0; i < s.length(); i++){
        if(charset[s[i]]){
            return false;
        }else{
            charset[s[i]] = true;
        }
    }
    return true;
}

// using a map to store whether we have seen characters or not
bool isUnique3(std::string& s){
    std::unordered_map<char, bool> umap = {};

    for(int i = 0; i < s.length(); i++){
        if(umap.find(s[i]) != umap.end()){
            return false;
        }else{
            umap[s[i]] = true;
        }
    }
    return true;
}

// sort the string first, and then we only need to iterate through it once
bool isUnique4(std::string& s){
    std::sort(s.begin(), s.end());

    for(int i = 0; i < s.length() - 1; i++){
        if(s[i] == s[i+1]){
            return false;
        }
    }
    return true;
}

int main(){
    std::string s1 = "hello world";
    std::string s2 = "abcdefg";
    std::cout << "1)" << s1 << ": " << isUnique1(s1) << "\n";
    std::cout << "1)" << s2 << ": " << isUnique1(s2) << "\n";
    std::cout << "2)" << s1 << ": " << isUnique2(s1) << "\n";
    std::cout << "2)" << s2 << ": " << isUnique2(s2) << "\n";
    std::cout << "3)" << s1 << ": " << isUnique3(s1) << "\n";
    std::cout << "3)" << s2 << ": " << isUnique3(s2) << "\n";
    std::cout << "4)" << s1 << ": " << isUnique4(s1) << "\n";
    std::cout << "4)" << s2 << ": " << isUnique4(s2) << "\n";
    return 1;
}
